### PR TITLE
Add a few test cases

### PR DESCRIPTION
**Pull request by xxdavid**
Discovered while testing Gradualizer on an Elixir repo.

### CODE REVIEW

File: test/known_problems/should_pass/elixir_list_first.erl [hunk 1]
```diff
@@ -0,0 +1,17 @@
+-module(elixir_list_first).
+
+-export([f/0]).
+
+% This is how List.first/1 is defined in Elixir
+-spec first([]) -> nil;
+           ([Elem, ...]) -> Elem.
+first([]) -> nil;
+first(Element) -> Element.
```
Comment (zuiderkwast) [inline test/known_problems/should_pass/elixir_list_first.erl:9]: Typo?

```suggestion
first([Element | _]) -> Element.
```

File: test/known_problems/should_pass/guard_negative_refinement_should_pass.erl [hunk 2]
```diff
@@ -0,0 +1,15 @@
+-module(guard_negative_refinement_should_pass).
```
Comment (zuiderkwast) [inline test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:1]: Negative? I believe `not_equal` is better when talking about the `/=` operator.

Btw, do we have tests for guards with `=/=`? And what about `==` and `=:=`?

File: test/known_problems/should_pass/guard_negative_refinement_should_pass.erl [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+-export([i/1]).
+
+-type state() :: #{value := integer() | nil}.
+
+-spec i(state()) -> integer().
+i(State) when map_get(value, State) /= nil ->
```
Comment (zuiderkwast) [inline test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:8]: This guard can fail. Shouldn't we have another clause for value nil? (Not that we can detect this non-exhaustiveness now, but the code makes more sense that way and maybe in the future we can check it.)

**Review by zuiderkwast — COMMENTED** [review]
Nice

File: test/known_problems/should_pass/elixir_list_first.erl (hunk 1 above)
Comment (xxdavid) [reply test/known_problems/should_pass/elixir_list_first.erl:9]: Yes, of course, thank you. 🙂

File: test/known_problems/should_pass/guard_negative_refinement_should_pass.erl (hunk 3 above)
Comment (xxdavid) [reply test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:8]: Good point, I've added a clause for `nil` (d3b5fa06).

Comment (xxdavid) [reply test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:8]: Fixed: 0c9307cc. I've also added `nil` cases for the other functions (2b2c5110).

File: test/known_problems/should_pass/guard_negative_refinement_should_pass.erl (hunk 2 above)
Comment (xxdavid) [reply test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:1]: Hmm, the *negative* part was meant like *type negation* (e.g. `(integer() | nil) \ nil`) but this is already contained in the *refinement*… So maybe it should be only `guard_refinement.erl`? There's one file that already tests guard refinements (including `=/=`, `==`, and `=:=`), `test/should_pass/refine_comparison.erl`, so maybe I should merge the files and put it there? What do you think?

Comment (zuiderkwast) [reply test/known_problems/should_pass/guard_negative_refinement_should_pass.erl:1]: Yeah... merge the two, or rename the other one, or both, or what you find appropriate. :) Now you have the full picture.

Comment (xxdavid) [reply test/known_problems/should_pass/guard_negative_refinement_should_pass.erl]: Okay. :) I merged it into the `test/should_pass/refine_comparison.erl` and created `test/known_problems/should_pass/refine_comparison_should_pass.erl` with the failing test. I've also squashed the fixups and force pushed.

**Review by zuiderkwast — APPROVED** [review]
LGTM

Btw, maybe you should consider patching your editor. :-) A decent editor automatically adds a newline at end-of-file.

**Comment by xxdavid** [comment]
Oh, thanks for noticing. Newlines added, editor configured (I though it was already like that but apparently not).

And thank you for the review. :)

**Pull request merged**